==========================
TESTS FOR THE matrix_mul MODULE
THIS CODE WILL BE COMPILED WITH:
python3 -m doctest -v ./tests/100-matrix_mul.txt
==========================

IMPORT THE MODULE
=================

    >>> matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul

TEST CASE: SUCCESS

PROD OF TWO MATRICES
matrix as a list of lists
consisting of integers and float values
(signed and unsigned)
====================

    >>> matrix_mul([[1, 2], [3, 4]], [[1, 2], [3, 4]])
    array([[ 7, 10],
           [15, 22]])

    >>> matrix_mul([[1.1, -2.2], [-5.5111111, 6]], [[2, 3], [4, -7]])
    array([[ -6.6      ,  18.7      ],
           [ 12.9777778, -58.5333333]])

TEST CASE: ERROR
================

NO ARGS
=======
    >>> matrix_mul()
    Traceback (most recent call last):
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'

PASSING ONLY ONE PARAMETER
==========================
    >>> matrix_mul([[6, 3, 2], [9, 2, 1 ]])
    Traceback (most recent call last):
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

NON LIST VALUES AS PARAMETERS
==================================
    >>> matrix_mul([[3, 6]], 9)
    Traceback (most recent call last):
    ValueError: Scalar operands are not allowed, use '*' instead

    >>> matrix_mul("hey young world", [[1, 2]])
    Traceback (most recent call last):
    ValueError: Scalar operands are not allowed, use '*' instead

    >>> matrix_mul(["Monty", "Python"], [[1, 2]])
    Traceback (most recent call last):
    ValueError: shapes (2,) and (1,2) not aligned: 2 (dim 0) != 1 (dim 0)

    >>> matrix_mul([[1, 2]], ["Monty", "Python"])
    Traceback (most recent call last):
    TypeError: invalid data type for einsum

EMPTY LIST
==========

    >>> matrix_mul([], [[9, 4]])
    Traceback (most recent call last):
    ValueError: shapes (0,) and (1,2) not aligned: 0 (dim 0) != 1 (dim 0)

    >>> matrix_mul([[9, 4]], [[]])
    Traceback (most recent call last):
    ValueError: shapes (1,2) and (1,0) not aligned: 2 (dim 1) != 1 (dim 0)

    >>> matrix_mul([[3, 6]], 9)
    Traceback (most recent call last):
    ValueError: Scalar operands are not allowed, use '*' instead

NON INT/FLOAT VALUES IN LIST
============================

    >>> matrix_mul([[1, 2], ["multiplied", 9]], [[9, 4], [2, 5]])
    Traceback (most recent call last):
    TypeError: invalid data type for einsum

    >>> matrix_mul([[9, 4], [2, 5]], [[1, 2], ["multiplied", 9]])
    Traceback (most recent call last):
    TypeError: invalid data type for einsum

IRREGULAR SIZED MATRIX
======================
    >>> matrix_mul([[1, 1], [2]], [[1, 1], [2, 2]])
    Traceback (most recent call last):
    ValueError: setting an array element with a sequence.

    >>> matrix_mul([[1, 1], [2, 2]], [[1], [2, 2]])
    Traceback (most recent call last):
    ValueError: setting an array element with a sequence.

MATRIX m_a IS BIGGER THAN MATRIX m_b
====================================
    >>> matrix_mul([[9, 4, 4], [2, 8, 2]], [[1, 2], [4, 9]])
    Traceback (most recent call last):
    ValueError: shapes (2,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)

